packages/kbn-esql-utils/src/utils/append_to_query.ts

```diff
@@ -5,9 +5,73 @@
  * in compliance with, at your election, the Elastic License 2.0 or the Server
  * Side Public License, v 1.
  */
+import { getAstAndSyntaxErrors } from '@kbn/esql-ast';
 
 // Append in a new line the appended text to take care of the case where the user adds a comment at the end of the query
 // in these cases a base query such as "from index // comment" will result in errors or wrong data if we don't append in a new line
 export function appendToESQLQuery(baseESQLQuery: string, appendedText: string): string {
   return `${baseESQLQuery}\n${appendedText}`;
 }
+
+export function appendWhereClauseToESQLQuery(
+  baseESQLQuery: string,
+  field: string,
+  value: unknown,
+  operation: '+' | '-',
+  fieldType?: string
+): string {
+  let operator = operation === '+' ? '==' : '!=';
+  let filterValue = typeof value === 'string' ? `"${value.replace(/"/g, '\\"')}"` : value;
+  // Adding the backticks here are they are needed for special char fields
+  let fieldName = `\`${field}\``;
+
+  // casting to string
+  // there are some field types such as the ip that need
+  // to cast in string first otherwise ES will fail
+  if (fieldType !== 'string' && fieldType !== 'number' && fieldType !== 'boolean') {
+    fieldName = `${fieldName}::string`;
+  }
+
+  // checking that the value is not null
+  // this is the existence filter
+  if (field === '_exists_') {
+    fieldName = `\`${String(value)}\``;
+    operator = ' is not null';
+    filterValue = '';
+  }
+
+  const { ast } = getAstAndSyntaxErrors(baseESQLQuery);
+
+  const lastCommandIsWhere = ast[ast.length - 1].name === 'where';
+  // if where command already exists in the end of the query:
+  // - we need to append with and if the filter doesnt't exist
+  // - we need to change the filter operator if the filter exists with different operator
+  // - we do nothing if the filter exists with the same operator
+  if (lastCommandIsWhere) {
+    const whereCommand = ast[ast.length - 1];
+    const whereAstText = whereCommand.text;
+    // the filter already exists in the where clause
+    if (whereAstText.includes(field) && whereAstText.includes(String(filterValue))) {
+      const pipesArray = baseESQLQuery.split('|');
+      const whereClause = pipesArray[pipesArray.length - 1];
+      const matches = whereClause.match(new RegExp(field + '(.*)' + String(filterValue)));
```

Comment:
Instead of re-parsing the query, I think we should rely on the AST here.

The current code breaks in a few scenarios

## False positive: detecting an existing filter

For example, the check at 54 is `true` for this scenario.
```
        appendWhereClauseToESQLQuery(
          `from kibana_sample_data_logs 
        | limit 10 
        | where CIDR_MATCH(ip1, "127.0.0.2/32", "127.0.0.3/32")`,
          'ip',
          '127.0.0.2/32',
          '-',
          'string'
        )
```

It then applies the regular expression and adds the operator in place of the first comma, resulting in an invalid query
```
from kibana_sample_data_logs·
            | limit 10·
            | where CIDR_MATCH(ip!= \"127.0.0.2/32\", \"127.0.0.3/32\")
```

## Replaces a nested boolean expression

In the following example, a boolean expression is successfully detected, but it isn't at the top level of the `WHERE` clause so it shouldn't be changed.
```
appendWhereClauseToESQLQuery(
          `from kibana_sample_data_logs | limit 10
          | where case(agent.keyword == "foo", 1, 0) == 1`,
          'agent.keyword',
          'foo',
          '-',
          'string'
        )
```

Results in

```
from kibana_sample_data_logs | limit 10
              | where case(agent.keyword != \"foo\", 1, 0) == 1
```

If we instead inspect the AST and find a compatible operator which has the `field` and `value` as its arguments, it will be a more robust existence check and you can know at what level the comparison is occurring (i.e. that it isn't nested somewhere).

Unfortunately, I guess we'll still have to rely on the final string manipulation until we have code that takes the AST back to a string.

Reply:
@drewdaemon this was a path I tried to take in the beginning but I decided against it as this needs to be a recursive function which thanx to our types mess is very difficult to create. It sounds to me a very complicated and possibly imperfomant way to find what you want while you can do it easier with a regex.

About your first example, this was a bug I fixed here and also added a test https://github.com/elastic/kibana/pull/181399/commits/a6ddb094a5490fe1f2255d6cb4961e0817ace9f5

About your second one I think is ok and also you cant easily detect it even with ast.

To prove you this you can check this commit https://github.com/elastic/kibana/pull/181399/commits/8c143964924051480ae366d37733fddbedfaf1b3. You can. see the complexity of the recursive function and our types mess. I would personally prefer to go with the previous implementation but I am happy to hear your thoughts on that.

Reply:
If you're okay with the second scenario not being covered, I think it's fair to stick with your current implementation for now. 

What I think this really highlights is the need for a good AST traversal API so that you don't have to worry about writing a complicated routine and can instead use existing code e.g. 

```ts
const nodes = findAll(whereNode, 'function', ['==', '!=']);
const matchingTopLevelClause = nodes.find(node => node.arguments[0].name == field && node.arguments[1].name == value && node.depth == 1);

...
```

Or maybe

```ts
const matcher = (node) => node.depth == 1
  && isFunctionNode(node) 
  && ['==', '!='].includes(node.name)
  && node.arguments[0].name == field
  && node.arguments[1].name == value;

const existingClause = findInAst(whereNode, matcher);

...
```

Reply:
I absolutely agree ++